etag.go

```diff
@@ -0,0 +1,47 @@
+package main
+
+import (
+	"crypto/sha1"
+	"encoding/binary"
+	"fmt"
+	"net/http"
+)
+
+// check whether client's ETag matches current response body.
+// - if the IMGPROXY_USE_ETAG env var is unset, this function always returns false
+// - if the IMGPROXY_USE_ETAG is set, the function calculates current ETag and compare it
+//   with another ETag value provided by client
+// Note that the calculated ETag is saved to outcoming response with "ETag" header.
+func isETagMatching(b []byte, po *processingOptions, rw *http.ResponseWriter, r *http.Request) bool {
+
+	if conf.ETagEnabled {
+
+		// calculate current ETag value using sha1 hashing function
+		currentEtagValue := calculateHashSumFor(b, po)
+		(*rw).Header().Set("ETag", currentEtagValue)
+		return currentEtagValue == r.Header.Get("If-None-Match")
+	}
+
+	return false
+}
+
+// function calculates the SHA checksum for the current image and current Processing Options.
+// Principal is very simple: if an original image is the same and PO are the same, then
+// the checksum must be always identical. But if PO has some different parameters, the
+// checksum must be different event if the orinal image matches
+func calculateHashSumFor(b []byte, po *processingOptions) string {
```

Comment:
I think this way to calc ETag is not perfectly correct.

First of all, it messes image data with config, this may cause collisions.
Then it writes `po` fields one by one. Someone can add a new field to `processingOptions`, and this code will become invalid.
And it uses `conf.Salt` which keeps being constant between restarts.

What should we do:
1) Calc SHA of image data. This way image footprint will have a constant length.
2) Make all fields of `processingOptions` public. This will allow providing the whole struct to `binary.Write`.
3) Generate random cache signature at each start of imgproxy. This will reset ETag in case use cached the imgproxy config.

ETag will be calculated like this: `sha(image_footprint + po + cacheSignature)`